geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala

```diff
@@ -0,0 +1,183 @@
+
+import java.util
+
+import com.google.common.cache.{CacheLoader, CacheBuilder}
+import kafka.admin.AdminUtils
+import kafka.utils.ZKStringSerializer
+import org.I0Itec.zkclient.ZkClient
+import org.I0Itec.zkclient.exception.ZkNodeExistsException
+import org.geotools.data.DataStore
+import org.geotools.feature.NameImpl
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.`type`.Name
+import org.opengis.feature.simple.SimpleFeatureType
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+trait KafkaDataStoreSchemaManager extends DataStore  {
+
+  protected def zookeepers: String
+  protected def zkPath: String
+  protected def partitions: Int
+  protected def replication: Int
+
+  //
+  // todo:  consistent exception handling across all methods
+  // todo   - catch and rethrow a known/consistent exception with helpful msg
+  //
+   override def createSchema(featureType: SimpleFeatureType): Unit = {
+
+    val kfc = KafkaFeatureConfig(featureType)  // this is guaranteed to have a topic
+
+    // build the schema node
+    val typeName = featureType.getTypeName
+    val schemaPath: String = getSchemaPath(typeName)
+    if (zkClient.exists(schemaPath)) {
+      throw new IllegalArgumentException(s"Type $typeName already exists")
+    }
+
+    val data = SimpleFeatureTypes.encodeType(featureType)
+    createZkNode(schemaPath, data)
+
+    // build the topic node
+    createZkNode(getTopicPath(typeName), kfc.topic)
+
+    // build the replay config node (optional)
+    // build the topic node
+    kfc.replayConfig.foreach(r => createZkNode(getReplayConfigPath(typeName), ReplayConfig.encode(r)))
+
+    //create the Kafka topic
+    if(!AdminUtils.topicExists(zkClient, kfc.topic))
+      AdminUtils.createTopic(zkClient, kfc.topic, partitions, replication)
+
+    // put it in the cache
+    schemaCache.put(typeName, kfc)
+
+  }
+
+  def getFeatureConfig(typeName: String) : KafkaFeatureConfig = schemaCache.get(typeName)
+
+  /** Extracts the prepared-for-live [[SimpleFeatureType]] which the given prepared-for-replay
+    * ``replayType`` is based on.
+    */
+  def getLiveFeatureType(replayType: SimpleFeatureType): Option[SimpleFeatureType] = {
+
+    Try {
+      KafkaDataStoreHelper.extractLiveTypeName(replayType).map(schemaCache.get(_).sft)
+    }.getOrElse(None)
+  }
+
+  override def getNames: util.List[Name] = zkClient.getChildren(zkPath).asScala.map(
+    name => new NameImpl(name) : Name).asJava
+
+
+  override def removeSchema(typeName: Name): Unit = removeSchema(typeName.getLocalPart)
+
+  override def removeSchema(typeName: String): Unit = {
+
+    // clean up cache and zookeeper resources
+    schemaCache.invalidate(typeName)
+    zkClient.deleteRecursive(getSchemaPath(typeName))
+
+    // todo: what about other remote caches - do we set up a kafka listener looking for deletes?
+    // todo: what about kafka - who has the responsibility to clean the kafka topic up?
```

Comment:
I replaced the TODOs with a reference to this ticket:
https://geomesa.atlassian.net/browse/GEOMESA-818